iotc-azrtos-sdk/src/iotconnect_di.c

```diff
@@ -0,0 +1,539 @@
+		// TODO: Remove this for obvious security reasons
+		printf("DDIM: Signing: %s + %s + hex_to_binary(\"%s\")\r\n", auth_rsp->rn, auth_rsp->cid, bin_to_hex(csr_bin, csr_len));
+		uint8_t sha_hash[TO_SHA256_HASHSIZE];
+		ret = _nx_crypto_sha256_digest_calculate(&ctx, sha_hash, NX_CRYPTO_HASH_SHA256);
+		if (ret) {
+			printf("DDIM: Unable to compute the hash\r\n");
+			ret = -9; goto cleanup;
+		}
+		uint8_t s_r_buffer[TO_SIGNATURE_SIZE];
+		if (ddim_interface->sign_hash(auth_interface_context, sha_hash, s_r_buffer)) {
+			printf("DDIM: Unable to sign the hash\r\n");
+			ret = -10; goto cleanup; // called function with print errors
+		}
+
+		// --- CONVERT SIG TO DER ----
+		size_t signature_size = 0;
+		uint8_t asn1_sig[TO_SIGNATURE_SIZE + 10]; // 2 bytes for header + 2 for each integer header + 4 bytes slack];
+		ret = sig_to_asn1_der(
+				asn1_sig, //
+				sizeof(asn1_sig),  //
+				&signature_size, //
+				&s_r_buffer[0], &s_r_buffer[TO_SIGNATURE_SIZE/2], TO_ECC_PRIV_KEYSIZE * 8 //
+				);
+		if (ret) {
+			printf("DDIM: Unable to convert signature S&R to der format\r\n");
+			ret = -11; goto cleanup;
+		}
+
+		// we can't reuse the work buffer below because we have two hex strings to return. We need to make a copy
+		char *sig_hex = bin_to_hex(asn1_sig, signature_size);
+		if (!sig_hex) {
+			printf("DDIM: Unable to convert signature to hex\r\n");
+			ret = -12; goto cleanup;
+		}
+
+		sign_req.sig = (char*)malloc(work_buffer_size); // work_buffer_size will be +1 for null
+		if (!sign_req.sig) {
+			printf("DDIM: Unable to allocate the signature string buffer\r\n");
+			ret = -13; goto cleanup;
+		}
+		memcpy(sign_req.sig, work_buffer, work_buffer_size);
+
+		sign_req.csr = bin_to_hex(csr_bin, csr_len);
+		if (!sign_req.csr) {
+			printf("DDIM: Unable to convert CSR to hex\r\n");
+			ret = -14; goto cleanup;
+		}
+
+		sign_req.fmt = "hex";
+
+		printf("DDIM: Requesting operational certificate.\r\n");
+		if (ddim_call_sign(azrtos_config, &sign_req, &sign_rsp)) {
+			ret = -15; goto cleanup; // called function with print errors
+		}
+	}
+
+	printf("DDIM: Received cert: %s\r\n", sign_rsp->cert);
+	hex_to_bin(sign_rsp->cert);
+	if (0 == work_buffer_size) {
+		printf("DDIM: Unable to convert cert to bin\r\n");
+		ret = -16; goto cleanup;
+	}
+
+	if(ddim_interface->store_operational_cert(auth_interface_context, &work_buffer[0], work_buffer_size)) {
+		ret = -17; goto cleanup; // called function with print errors
+	}
+
+	ack_req.cid = auth_rsp->cid;
+	printf("DDIM: Acknowledging cert: %s\r\n", sign_rsp->cert);
+	if (ddim_call_ack(azrtos_config, &ack_req, &ack_rsp)) {
+		ret = -18; goto cleanup; // called function with print errors
+	}
+
+	// pass through on success with ret = 0;
+	cleanup:
+	if (auth_rsp) {
+		iotcl_ddim_free_auth_response(auth_rsp);
+	}
+	if (sign_rsp) {
+		iotcl_ddim_free_sign_response(sign_rsp);
+	}
+	if (ack_rsp) {
+		iotcl_ddim_free_ack_response(sign_rsp);
+	}
+	if (sign_req.csr) {
+		free(sign_req.sig); // the malloc'd copy
```

Comment:
Looks odd -- test doesn't match what's being free'd

Reply:
we should check for .sig instead of .csr

Reply:
fixed.